core/outbox.go

```diff
@@ -0,0 +1,147 @@
+package core
+
+import (
+	"context"
+	"sync"
+
+	"github.com/filecoin-project/go-filecoin/abi"
+	"github.com/filecoin-project/go-filecoin/actor"
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/consensus"
+	"github.com/filecoin-project/go-filecoin/metrics"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/ipfs/go-cid"
+	"github.com/pkg/errors"
+)
+
+// Topic is the network pubsub topic identifier on which new messages are announced.
+const Topic = "/fil/msgs"
+
+// Outbox validates and marshals messages for sending and maintains the outbound message queue.
+type Outbox struct {
+	// Signs messages
+	signer types.Signer
+	// Validates messages before sending them.
+	validator consensus.SignedMessageValidator
+	// Holds messages sent from this node but not yet mined.
+	queue *MessageQueue
+	// Invoked to publish a message to the network
+	publisher publisher
+
+	chains chainProvider
+	actors actorProvider
+
+	// Protects the "next nonce" calculation to avoid collisions.
+	nonceLock sync.Mutex
+}
+
+type chainProvider interface {
+	GetHead() types.SortedCidSet
+	GetTipSet(tsKey types.SortedCidSet) (*types.TipSet, error)
+}
+
+type actorProvider interface {
+	// GetActor returns the actor state defined by the chain up to some tipset
+	GetActor(ctx context.Context, tipset types.SortedCidSet, addr address.Address) (*actor.Actor, error)
+}
+
+type publisher interface {
+	Publish(ctx context.Context, message *types.SignedMessage, height uint64) error
+}
+
+var msgSendErrCt = metrics.NewInt64Counter("message_sender_error", "Number of errors encountered while sending a message")
+
+// NewOutbox creates a new outbox
+func NewOutbox(signer types.Signer, validator consensus.SignedMessageValidator, queue *MessageQueue,
+	publisher publisher, chains chainProvider, actors actorProvider) *Outbox {
+	return &Outbox{
+		signer:    signer,
+		validator: validator,
+		queue:     queue,
+		publisher: publisher,
+		chains:    chains,
+		actors:    actors,
+	}
+}
+
+// Send marshals and sends a message, retaining it in the outbound message queue.
+func (ob *Outbox) Send(ctx context.Context, from, to address.Address, value *types.AttoFIL,
+	gasPrice types.AttoFIL, gasLimit types.GasUnits, method string, params ...interface{}) (out cid.Cid, err error) {
+	defer func() {
+		if err != nil {
+			msgSendErrCt.Inc(ctx, 1)
```

Comment:
If this is only executed in the event of no errors, why not just put it on line 122 and avoid the added complexity of the defer statement at all?

Reply:
This is executed only when there *is* an error. It can't go at the end due to the unstructured early return statements throughout. (I went through the same thought process).